### PR TITLE
Cleaning up OSize and exposing more knobs for code size tuning

### PR DESCRIPTION
**Pull request by rajbarik**
Improved Osize: 
(1) Use MinSize instead of OptForSize LLVM annotation; 
(2) Expose UnrollThreshold and Inline_Benefit_Reduction_Factor in the command line for automatic tuning of inlining and unrolling for code size; 
(3) Modify PassManager to enable Osize specific passes instead of current approach that checks for Osize inside individual passes

### CODE REVIEW

**Review by aschwaighofer — CHANGES_REQUESTED** [review]
@rajbarik Thank you for improvements.

Can you rebase to a recent master?
Can you split the following functionality into two commits:

(1) Use MinSize instead of OptForSize LLVM annotation;
(2) Expose UnrollThreshold and Inline_Benefit_Reduction_Factor in the command line for automatic tuning of inlining and unrolling for code size;

We intend to add a function level attribute to enable "-O" like optimization when compiling at -Osize for a function. This would not work with a statically modified pass pipeline.

(3) Modify PassManager to enable Osize specific passes instead of current approach that checks for Osize inside individual passes


Code conventions:
http://llvm.org/docs/CodingStandards.html

A good way to following the without much effort is to use clang-format (it has bindings for emacs/vim/git)

https://github.com/apple/swift-clang/tree/stable/tools/clang-format

File: lib/SILOptimizer/LoopTransforms/LoopUnroll.cpp [hunk 1]
```diff
@@ -29,7 +29,7 @@ using namespace swift::PatternMatch;
 using llvm::DenseMap;
 using llvm::MapVector;
 
-static const uint64_t SILLoopUnrollThreshold = 250;
+//static const uint64_t SILLoopUnrollThreshold = 250;
```
Comment (aschwaighofer) [inline lib/SILOptimizer/LoopTransforms/LoopUnroll.cpp]: Remove this line?

File: lib/SILOptimizer/LoopTransforms/LoopUnroll.cpp [hunk 2]
```diff
@@ -187,6 +187,8 @@ static bool canAndShouldUnrollLoop(SILLoop *Loop, uint64_t TripCount) {
   // It is used to estimate the cost of the callee
   // inside a loop.
   const uint64_t InsnsPerBB = 4;
+  // Threshold of unrolling
+  const uint64_t SILLoopUnrollThreshold = Loop->getBlocks().empty() ? 0 : (Loop->getBlocks())[0]->getParent()->getModule().getOptions().UnrollThreshold;
```
Comment (aschwaighofer) [inline lib/SILOptimizer/LoopTransforms/LoopUnroll.cpp]: break the line at 80 chars.

File: include/swift/AST/SILOptions.h [hunk 3]
```diff
@@ -32,6 +32,12 @@ class SILOptions {
   /// Controls the aggressiveness of the performance inliner.
   int InlineThreshold = -1;
 
+  /// Controls the aggressiveness of the performance inliner for Osize
```
Comment (aschwaighofer) [inline include/swift/AST/SILOptions.h]: Can you add a dot at the end of the line.

File: include/swift/AST/SILOptions.h [hunk 4]
```diff
@@ -32,6 +32,12 @@ class SILOptions {
...
+  int CallerBaseBenefitReductionFactor = 2;
+
+  /// Controls the aggressiveness of the loop unroller
```
Comment (aschwaighofer) [inline include/swift/AST/SILOptions.h]: Can you add a dot at the end of the line.

**Comment by rajbarik** [comment]
I created this: https://github.com/apple/swift/pull/12649 for part 1. Not sure if you still need part 2 that disables the optimizations at pass-pipeline level. Let me know.

**Pull request closed without merging**